src/stream/src/executor/dispatch.rs

```diff
@@ -161,6 +154,30 @@ impl DispatchExecutorInner {
     }
 
     async fn dispatch(&mut self, msg: MessageBatch) -> StreamResult<()> {
+        macro_rules! await_with_metrics {
+            ($fut:expr, $metrics:expr, $interval:expr, $start_time:expr) => {{
+                // First tick completes immediately: https://docs.rs/tokio/1.47.1/tokio/time/fn.interval.html.
+                $interval.tick().await;
+
+                loop {
+                    tokio::select! {
+                        biased;
+                        res = &mut $fut => {
+                            res?;
+                            let ns = $start_time.elapsed().as_nanos() as u64;
+                            $metrics.inc_by(ns);
+                            break;
+                        }
+                        _ = $interval.tick() => {
+                            $start_time = Instant::now();
+                            $metrics.inc_by(Duration::from_secs(15).as_nanos() as u64);
```

Comment:
```suggestion
                            $metrics.inc_by($interval.period().as_nanos() as u64);
```

Reply:
Oh, I think we can maintain `$interval` directly inside the macro expanded macro?